File: ts/src/baseclient.ts
import * as $tea from '@alicloud/tea-typescript';
import uuid from 'uuid/v4';
import { stringify } from 'querystring';
import { createHmac,createHash } from 'crypto';

const filterKey = ['x-ca-signature', 'x-ca-signature-headers', 'accept', 'content-md5', 'content-type', 'date', 'host', 'token', 'reflect']

export default class BaseClient {

  _regionId: string
  _protocol: string
  _domain: string
  _token: string
  _appKey: string
  _appSecret: string
  _stage: string
  _readTimeout: number
  _connectTimeout: number
  _localAddr: string
  _httpProxy: string
  _httpsProxy: string
  _noProxy: string
  _maxIdleConns: number

  constructor(config: { [key: string]: any }) {
    this._protocol = config['protocol'] || 'http';
    this._regionId = config['regionId'] || '';
    this._domain = config['domain'] || '';
    this._token = config['token'] || '';
    this._appKey = config['appKey'] || '';
    this._appSecret = config['appSecret'] || '';
    this._readTimeout = config['readTimeout'] || 0;
    this._connectTimeout = config['connectTimeout'] || 0;
    this._localAddr = config['localAddr'] || '';
    this._httpProxy = config['httpProxy'] || '';
    this._httpsProxy = config['httpsProxy'] || '';
    this._noProxy = config['noProxy'] || '';
    this._maxIdleConns = config['maxIdleConns'] || 0;
  }

  _getHost(): string {
    return this._domain;
  }

  _isFail(resp: $tea.Response): boolean {
    return resp.statusCode < 200 || resp.statusCode >= 300;
  }

  _defaultNumber(number: number, defaultNum: number): number {
    if (!number) {
      return defaultNum;
    }
    return number;
  }

  _toForm(form: { [key: string]: any }): string {
    return stringify(form);
  }

  _toJSONString(body: any): string {
    return JSON.stringify(body);
  }

  _getUUID(): string {
    return uuid();
  }

  _default(value: string, defaultVal: string): string {
    if (!value) {
      return defaultVal;
    }
    return value;
  }

  _getContentMD5(body: { [key: string]: any }): string {
    const bodyJson = JSON.stringify(body);
    const hash = createHash('md5');
    hash.update(bodyJson);
    return hash.digest('hex');
  }

  _buildUrl(request: $tea.Request): string{
    let url = request.pathname;
    let queryKeys = Object.keys(request.query);
    queryKeys.sort();
    if (queryKeys.length > 0) {
      url += '?';
    }

    queryKeys.forEach((key) => {
      if (!url.endsWith('?')) {
        url += '&';
      }
      if (typeof request.query[key] === 'undefined') {
        url += encodeURIComponent(key);
      } else {
        url += `${encodeURIComponent(key)}=${encodeURIComponent(request.query[key])}`;
      }
    });
    return url;
  }

  _getSignedHeader(request: $tea.Request): string {
    let signedHeader = '';
    let signedHeaderKeys = '';
    let headerKeys = Object.keys(request.headers);
    headerKeys.sort();

    headerKeys.forEach((key) => {
      if (~filterKey.indexOf(key)) {
        return;
      }
      signedHeaderKeys += `${key},`;
      signedHeader += `${key}:${request.headers[key]}\n`;

    });
    request.headers["x-ca-signature-headers"] = signedHeaderKeys.slice(0, -1);
    return signedHeader.slice(0, -1);
  }

  _getSignature(request: $tea.Request): string {
    const signedHeader = this._getSignedHeader(request);
    const url = this._buildUrl(request);
    // const signStr = `${request.method}\n${accept}\n${contentMd5}\n${contentType}\n${date}\n${signedHeader}\n${url}`;
    const hmac = createHmac('sha256', this._appSecret);
    hmac.update(request.method + '\n');
    hmac.update((request.headers['accept'] || '') + '\n');
    hmac.update((request.headers['content-md5'] || '') + '\n');
    hmac.update((request.headers['content-type'] || '') + '\n');
    hmac.update(request.headers['date'] + '\n');
    hmac.update(signedHeader + '\n');
    hmac.update(url);
    return hmac.digest('base64');
  }

  _isStatusCode(response: $tea.Response, code: number): boolean {
    return response.statusCode === code;
  }

  async _readAsJSON(request: $tea.Response): Promise<{ [key: string]: any }> {
    let body = await request.readBytes();
    body = JSON.parse(body.toString());
    return body;
  }

  _getDate(): string {
    let now = new Date();
    return now.toUTCString();
  }

  _getTimestamp(): string {
    return Date.now().toString();
  }

  _equal(realStr: string , defaultStr: string): boolean{
    return realStr === defaultStr;
  }

  _notNull(obj: { [key: string]: any }): boolean{
    if (typeof obj === 'undefined' || obj == null) {
      return false;
    }
    return Object.keys(obj).length > 0
  }

  _toQuery(query: { [key: string]: any } , prefix: string = ''): { [key: string]: string } {
    let ret: { [key: string]: string } = {};
    if (!this._notNull(query)) {
      return ret;
    }
    Object.keys(query).forEach(key => {
      if (typeof query[key] === 'undefined' || query[key] == null) {
        return;
      }
      const newKey = prefix + key;
      if (query[key] instanceof Object) {
        const tmp = this._toQuery(query[key], newKey + '.');
        ret = {
          ...ret,
          ...tmp,
        };
        return;
      }
      ret[newKey] = query[key].toString();
    })
    return ret;
  }

}

File: java/src/main/java/com/apigateway/baseclient/BaseClient.java
package com.apigateway.baseclient;

import com.aliyun.tea.TeaRequest;
import com.aliyun.tea.TeaResponse;
import com.aliyun.tea.utils.StringUtils;
import com.google.gson.Gson;

import javax.crypto.Mac;
import javax.crypto.spec.SecretKeySpec;
import java.io.IOException;
import java.io.UnsupportedEncodingException;
import java.net.URLEncoder;
import java.security.InvalidKeyException;
import java.security.MessageDigest;
import java.security.NoSuchAlgorithmException;
import java.text.SimpleDateFormat;
import java.util.*;

public class BaseClient {
    protected String _appKey;
    protected String _stage;
    protected String _token;
    protected String _domain;
    protected String _appSecret;

    protected long _readTimeout;
    protected long _connectTimeout;
    protected String _localAddr;
    protected String _httpProxy;
    protected String _httpsProxy;
    protected String _noProxy;
    protected long _maxIdleConns;
    protected String _protocol;

    private static final String[] moveList = new String[]{
            "x-ca-signature", "x-ca-signature-headers", "accept", "content-md5", "content-type", "date", "host", "token"};


    public BaseClient(Map<String, Object> map) {
        this._appKey = (String) map.get("appKey");
        this._stage = (String) map.get("stage");
        this._token = (String) map.get("token");
        this._domain = (String) map.get("domain");
        this._appSecret = (String) map.get("appSecret");
        this._protocol = map.get("protocol") == null ? "http" : (String) map.get("protocol");
        this._noProxy = (String) map.get("noProxy");
        this._httpsProxy = (String) map.get("httpsProxy");
        this._httpProxy = (String) map.get("httpProxy");
        this._localAddr = (String) map.get("localAddr");
        this._readTimeout = map.get("readTimeout") == null ? 10000 : Long.parseLong(String.valueOf(map.get("readTimeout")));
        this._connectTimeout = map.get("connectTimeout") == null ? 5000 : Long.parseLong(String.valueOf(map.get("connectTimeout")));
        this._maxIdleConns = map.get("maxIdleConns") == null ? 0 : Long.parseLong(String.valueOf(map.get("maxIdleConns")));
    }

    protected Number _defaultNumber(Integer maxAttempts, long defaultNumber) {
        if (maxAttempts != null && maxAttempts >= 0) {
            return maxAttempts;
        }
        return defaultNumber;
    }


    protected String _default(String backoffPolicy, String no) {
        return StringUtils.isEmpty(backoffPolicy) ? no : backoffPolicy;
    }

    protected String _getHost() {
        return this._domain;
    }

    protected String _getDate() {
        SimpleDateFormat df = new SimpleDateFormat("EEE, dd MMM yyyy HH:mm:ss z", Locale.US);
        df.setTimeZone(TimeZone.getTimeZone("GMT"));
        return df.format(new Date());
    }

    protected String _getTimestamp() {
        return String.valueOf(System.currentTimeMillis());
    }

    protected String _getUUID() {
        return UUID.randomUUID().toString();
    }

    protected String _toJSONString(Map<String, Object> map) {
        return new Gson().toJson(map);
    }

    protected String _getContentMD5(Map<String, Object> map) throws NoSuchAlgorithmException, IOException {
        if (null == map || map.size() <= 0) {
            return "";
        }
        String bodyStr = new Gson().toJson(map);
        MessageDigest md = MessageDigest.getInstance("MD5");
        byte[] result = md.digest(bodyStr.getBytes(TeaRequest.URL_ENCODING));
        return Base64.getEncoder().encodeToString(result);
    }

    protected String _getSignature(TeaRequest request) throws NoSuchAlgorithmException, InvalidKeyException, UnsupportedEncodingException {
        String httpMethod = request.method;
        String accept = request.headers.get("accept");
        String contentMd5 = request.headers.get("content-md5") == null ? "" : request.headers.get("content-md5");
        String contentType = request.headers.get("content-type") == null ? "" : request.headers.get("content-type");
        String date = request.headers.get("date");
        String header = getSignHeader(request);
        String url = getUrl(request);
        String stringToSign = httpMethod + "\n" + accept + "\n" + contentMd5 + "\n" + contentType + "\n" +
                date + "\n" + header + "\n" + url;
        Mac hmacSha256 = Mac.getInstance("HmacSHA256");
        byte[] keyBytes = this._appSecret.getBytes(TeaRequest.URL_ENCODING);
        hmacSha256.init(new SecretKeySpec(keyBytes, 0, keyBytes.length, "HmacSHA256"));
        String sign = new String(Base64.getEncoder().encode(hmacSha256.doFinal(stringToSign.getBytes(TeaRequest.URL_ENCODING))));
        return sign;
    }

    private String getSignHeader(TeaRequest request) {
        if (request.headers.size() <= 0) {
            return "";
        }
        Map<String, String> map = new HashMap<>();
        map.putAll(request.headers);
        for (String moveName : moveList) {
            map.remove(moveName);
        }
        Set<String> keySet = map.keySet();
        String[] keys = keySet.toArray(new String[]{});
        Arrays.sort(keys);
        StringBuilder stringBuilder = new StringBuilder();
        StringBuilder headerKeys = new StringBuilder();
        for (String key : keys) {
            headerKeys.append(key).append(",");
            stringBuilder.append(key).append(":").append(map.get(key) == null ? "" : map.get(key)).append("\n");
        }
        if (headerKeys.length() >= 1) {
            headerKeys.deleteCharAt(headerKeys.length() - 1);
        }
        if (stringBuilder.length() >= 1) {
            stringBuilder.deleteCharAt(stringBuilder.length() - 1);
        }
        request.headers.put("x-ca-signature-headers", headerKeys.toString());
        return stringBuilder.toString();
    }

    private String getUrl(TeaRequest request) {
        String pathName = request.pathname;
        Map<String, String> map = new HashMap<>();
        map.putAll(request.query);
        String[] keys = map.keySet().toArray(new String[]{});
        StringBuilder stringBuilder = new StringBuilder(pathName);
        if (map.size() > 0) {
            stringBuilder.append("?");
            for (String key : keys) {
                stringBuilder.append(key).append("=").append(map.get(key)).append("&");
            }
            stringBuilder.deleteCharAt(stringBuilder.length() - 1);
        }
        return stringBuilder.toString();
    }

    protected boolean _isStatusCode(TeaResponse response_, int code) {
        return response_.statusCode == code;
    }

    protected Map<String, Object> _readAsJSON(TeaResponse response_) throws IOException {
        String body = response_.getResponseBody();
        Gson gson = new Gson();
        Map<String, Object> map = gson.fromJson(body, Map.class);
        return map;
    }

    protected boolean _equal(String bodystr, String type) {
        if (StringUtils.isEmpty(bodystr) || StringUtils.isEmpty(type)) {
            return false;
        }
        return type.equals(bodystr);
    }

    protected String _toForm(Map<String, ?> map) throws UnsupportedEncodingException {
        if (null == map) {
            return "";
        }
        StringBuilder result = new StringBuilder();
        boolean first = true;
        for (Map.Entry<String, ?> entry : map.entrySet()) {
            if (entry.getValue() == null) {
                continue;
            }
            if (first) {
                first = false;
            } else {
                result.append("&");
            }
            result.append(URLEncoder.encode(entry.getKey(), "UTF-8"));
            result.append("=");
            result.append(URLEncoder.encode(String.valueOf(entry.getValue()), "UTF-8"));
        }
        return result.toString();
    }

    protected boolean _notNull(Map<String, ?> map) {
        if (null == map) {
            return false;
        }
        return map.size() > 0;
    }

    protected Map<String, String> _toQuery(Map<String, ?> map) {
        Map<String, String> query = new HashMap<>();
        if (null == map) {
            return query;
        }
        for (Map.Entry<String, ?> entry : map.entrySet()) {
            if (entry.getValue() == null) {
                continue;
            }
            query.put(entry.getKey(), String.valueOf(entry.getValue()));
        }
        return query;
    }

    protected boolean _isFail(TeaResponse response) {
        if (null == response) {
            return false;
        }
        if (200 > response.statusCode || 300 <= response.statusCode) {
            return true;
        }
        return false;
    }
}
